File: board.py
from pieces import Piece, General
from typing import List, Optional, Callable
from supports import Vector, Color, textcolors, FULLBOARD_AREA, RED_PALACE_AREA, BLACK_PALACE_AREA, EvaluateSet, Square_status

class Square:
    def __init__(self, position: Vector):
        self.position = position
        self.piece: Optional['Piece'] = None
        self.state = Square_status.neutral
        self.red_attacks: int = 0
        self.black_attacks: int = 0

    def reset_attacks(self):
        self.red_attacks = 0
        self.black_attacks = 0

    def add_attack(self, color: Color):
        if color == Color.RED:
            self.red_attacks += 1
        else:
            self.black_attacks += 1
        self.update_state()
    
    def remove_attack(self, color: Color):
        if color == Color.RED:
            self.red_attacks -= 1
        else:
            self.black_attacks -= 1
        self.update_state()

    def update_state(self):
        if self.red_attacks > self.black_attacks:
            self.state = Square_status.red
        elif self.black_attacks > self.red_attacks:
            self.state = Square_status.black
        else:
            self.state = Square_status.neutral

    def get_state(self):
        return self.state

    def get_piece(self):
        return self.piece

    def remove_piece(self):
        self.piece = None

    def set_piece(self, piece: Piece):
        old_piece = self.piece
        self.piece = piece
        piece.set_position(self.position)
        return old_piece

class Board:
    def __init__(self, debug=False):
        self.squares = [[Square(Vector(x, y)) for y in range(10)] for x in range(9)]
        self.evaluation = 0
        self.reds = []
        self.blacks = []
        self.debug = debug

    def add_piece(self, piece: Piece):
        position = piece.get_position()
        square = self.get_square(position)
        if square and square.get_piece():
            raise ValueError(f"Square {position} is already occupied by {square.piece}")
        else:
            square.set_piece(piece)
        if piece.color == Color.RED:
            self.reds.append(piece)
        else:
            self.blacks.append(piece)

    def remove_piece(self, piece: Optional['Piece']):
        if piece.color == Color.RED:
            self.reds.remove(piece)
        else:
            self.blacks.remove(piece)

    def get_square(self, position: Vector, area:Optional[tuple]=None):
        """
        params: area: part of the board to check if square is in this area
                None means area checking is not needed 
        """
        if area and not(position.in_area(area)):
            return None
        return self.squares[position.x][position.y]

    # returns the square of piece, giving check
    # if is not in check, returns None
    def is_in_check(self, color: Color) -> Optional[Square]:
        opponent_pieces = self.reds if color == Color.BLACK else self.blacks
        palace = RED_PALACE_AREA if color == Color.RED else BLACK_PALACE_AREA
        attackers = [attacker for attacker in opponent_pieces if attacker.is_attacker()]
        for attacker in attackers:
            attackers_valid_moves = self.get_piece_valid_moves(attacker, check=False)
            for move in attackers_valid_moves:
                square = self.get_square(move, area=palace)
                if square and square.piece and isinstance(square.piece, General):
                    return attacker.get_position()
        return None
    
    def get_attackers(self, color: Color) -> List[Piece]:
        return [piece for piece in (self.reds if color == Color.RED else self.blacks) if piece.is_attacker()]
    
    def move_piece(self, from_pos: Vector, to_pos: Vector, set_instead : Optional['Piece'] = None):
        old_square = self.squares[from_pos.getX()][from_pos.getY()]
        the_piece = old_square.get_piece()
        if the_piece is None:
            raise ValueError("There is no piece on " + from_pos)
        the_piece.set_position(to_pos)
        taken = self.squares[to_pos.getX()][to_pos.getY()].set_piece(the_piece)
        if taken is not None:
            self.remove_piece(taken)

        old_square.remove_piece()    

        if set_instead is not None:
            self.add_piece(set_instead)
        return taken

    def ghost_test(self, from_pos:Vector, to_pos:Vector, func: Callable):
        taken = self.move_piece(from_pos, to_pos)
        result = func(self)
        self.move_piece(to_pos, from_pos, taken)
        return result

    def get_piece_valid_moves(self, piece: Piece, check=True):
        valid_moves = []

        if piece.can_modify() and (not piece.is_modified()):
            if (piece.position.getY() > 4 if (piece.get_color() == Color.RED) else piece.get_position().getY() < 5):
                piece.modify()
        
        move = piece.get_move()

        for direction in move.get_directions():
            
            if move.is_bigstep() and self.has_piece(piece.get_position() + direction // 2):
                continue

            has_screen = False
            position = piece.get_position()
            for step in range(piece.get_max_steps()):
                position += direction
                target_square = self.get_square(position, area=piece.get_area())
                if target_square is None:
                    break
                if target_square.get_piece() is not None:
                    if (not piece.need_screen()) or has_screen:
                        if target_square.get_piece().get_color() != piece.get_color():
                            if check:
                                threat = self.ghost_test(piece.get_position(), position, lambda xxx: xxx.is_in_check(piece.get_color()))
                                if threat is not None:
                                    if self.debug:
                                        print(f"Move by {piece} to {position} causes mate by {threat}")
                                    break
                            valid_moves.append(position)
                        break
                    else:
                        has_screen = True
                elif not has_screen:
                    if check:
                        threat = self.ghost_test(piece.get_position(), position, lambda xxx: xxx.is_in_check(piece.get_color()))
                        if threat is not None:
                            if self.debug:
                                print(f"Move by {piece} to {position} causes mate by {threat}")
                            break
                    valid_moves.append(position)

        if piece.can_modify() and piece.is_modified():
            piece.set_default()

        return valid_moves


    def has_piece(self, position: Vector) -> bool:
        sq = self.get_square(position, area=FULLBOARD_AREA)
        return bool(sq) and bool(sq.get_piece())

    def evaluate(self, eval_set: EvaluateSet, describe=False) -> float:
        total_value = 0
        self.reset_control_state()
        value_multiplier, attack_bonus, mobility_multiplier, control_multiplier = eval_set.get()
        for i,team in enumerate((self.reds, self.blacks)):
            team_values = team_attack_bonus = team_mobility = 0
            for piece in team:
                team_values += piece.get_value() * value_multiplier
                team_attack_bonus += attack_bonus if piece.is_attacker() else 0
                vds = self.get_piece_valid_moves(piece, check=False)
                team_mobility += len(vds)* mobility_multiplier
                self.update_control_state(piece.get_color(), vds)
            if describe:
                print(f"BOARD: {piece.get_color().name} values = {team_values}")
                print(f"BOARD: {piece.get_color().name} attack bonus = {team_attack_bonus}")
                print(f"BOARD: {piece.get_color().name} mobility = {team_mobility}")
            total_value += (team_values + team_attack_bonus + team_mobility) * ((-1) if i else 1)
        for rows in self.squares:
            for square in rows:
                total_value += square.get_state().value * (control_multiplier if square.get_piece() is None else (control_multiplier+attack_bonus))
        return total_value

    def update_evaluation(self):
        self.evaluation = self.evaluate(self.eval_set)
        print(f"BOARD: Evaluation = {self.evaluation}")
        return self.evaluation

    def is_mated(self):
        return abs(self.evaluation) >= 100

    def update_control_state(self, color:Color, attacks: List[Vector]):
        for attack in attacks:
            square = self.get_square(attack)
            square.add_attack(color)

    def reset_control_state(self):
        for row in self.squares:
            for square in row:
                square.reset_attacks()

    def get_reds(self):
        return self.reds

    def get_blacks(self):
        return self.blacks

    def print_pieces(self):
        print("RED PIECES:")
        for piece in self.reds:
            print(piece)
        print("BLACK PIECES:")
        for piece in self.blacks:
            print(piece)

    def print_visual(self, prespective: Color = Color.RED):
        line = " ╔"
        for i in range(9):
            line += "═╤═"
        line += "╗"
        print(line)
        
        for i in range(10):
            vert = i if prespective==Color.BLACK else (9-i)
            line = str(vert)
            line += "╟"
            for j in range(9):
                print_piece = self.squares[j][vert].get_piece()
                if i == 4:
                    between = "┴"
                elif i == 5:
                    between = "┬"
                else:
                    between = "┼"
                if print_piece:
                    line += " "
                    line += textcolors.red if print_piece.get_color()==Color.RED else textcolors.green
                    line += print_piece.get_name() + textcolors.endc + " "
                else:
                    line += "─" + between + "─"
            line += "╢"
            print(line)
        line = " ╚"
        for i in range(9):
            line += "═╧═"
        line += "╝"
        print(line)
        line = "   "
        for i in range(9):
            line += str(i) + "  "
        print(line)
        print()
        

File: game_logic/new_clang/clang/board.cpp
#ifndef board_cpp
#define board_cpp

#include "board.h"
#include <iostream>
#include <vector>
#include <memory>
#include <functional>
#include <algorithm>
#include <cmath>
#include <tuple>
#include <unordered_set>
#include "pieces.h"
#include <optional>

void Square::update_state() {
    if (red_attacks > black_attacks) {
        state = Square_status::red;
    } else if (black_attacks > red_attacks) {
        state = Square_status::black;
    } else {
        state = Square_status::neutral;
    }
}

std::unique_ptr<Piece> Square::set_piece(std::unique_ptr<Piece> new_piece) {
    auto old_piece = piece;
    piece = new_piece;
    if (new_piece) {
        new_piece->set_position(position);
    }
    return old_piece;
}

void Square::add_attack(Color color) {
    if (color == Color::RED) {
        ++red_attacks;
    } else {
        ++black_attacks;
    }
    update_state();
}

void Square::remove_attack(Color color) {
    if (color == Color::RED) {
        red_attacks--;
    } else {
        black_attacks--;
    }
    update_state();
}

void Board::add_piece(std::unique_ptr<Piece> piece) {
    Vector position = piece->get_position();
    Square& square = get_square(position);
    
    if (square.get_piece()) {
        throw std::runtime_error("Square is already occupied");
    }
    
    square.set_piece(piece);
    
    if (piece->get_color() == Color::RED) {
        reds.push_back(piece);
    } else {
        blacks.push_back(piece);
    }
}

void Board::remove_piece(std::unique_ptr<Piece> piece) {
    auto& collection = (piece->get_color() == Color::RED) ? reds : blacks;
    auto it = std::find(collection.begin(), collection.end(), piece);
    if (it != collection.end()) {
        collection.erase(it);
    }
}

Square& Board::get_square(const Vector& position, 
                    const std::optional<std::pair<Vector, Vector>>& area = std::nullopt) {
    if (area) {
        if (!position.in_area(*area)) {
            throw std::out_of_range("Position outside specified area");
        }
    }
    
    if (position.x < 0 || position.x >= 9 || position.y < 0 || position.y >= 10) {
        throw std::out_of_range("Invalid board position");
    }
    
    return squares[position.x][position.y];
}

bool Board::empty_between_general(int x, int y1, int y2) const {
    if (y1 > y2) std::swap(y1, y2);
    for (int y = y1 + 1; y < y2; ++y) {
        if (squares[x][y].get_piece()) {
            return false;
        }
    }
    return true;
}

void Board::get_attackers(Color color) const {
    auto& collection = (color == Color::RED) ? reds : blacks;
    std::vector<std::unique_ptr<Piece>> attackers;
    
    for (auto& piece : collection) {
        if (piece->is_attacker()) {
            attackers.push_back(piece);
        }
    }
    // return attackers;
}

void Board::move_piece(std::unique_ptr<Piece> piece, Vector dest){
    Vector from_pos = piece->get_position();
    piece->set_position(dest);
    for (auto& p : pieces){
        '''
            Цикл по всем фигурам, чтобы проверить, как на них повлиял сделанный ход
        '''
        if (p->get_position == dest) { // if the piece has taken
            remove_piece(p);
        }else{
            int i = p->block_direction(from_pos, p->get_color());
            '''
                Для того чтобы продлить за точку границы нужно найти новую.
                Для этого нужно пробежать по остальным фигурам.
            '''
            int min_dist = 10 // BOARD SIZE 
            for (auto& new_p : pieces) {
                if (new_p->is_reachable_for(p)){
                    Vector dist_v = p->get_position() - new_p->get_position();
                    int dist = dist_v.length()
                    if dist < min_dist{
                        
                    }
                }
            }
            p->block_direction(dest, piece->get_color());
        }
    }
    for (auto& piece : blacks){
        piece->get_blocked_direction(dest, piece->get_color());
    }
    
}

std::unique_ptr<Piece> Board::move_piece(const Vector& from_pos, const Vector& to_pos,
                                    std::unique_ptr<Piece> set_instead = nullptr,
                                    bool returning = false) {
    Square& old_square = get_square(from_pos);
    std::unique_ptr<Piece> the_piece = old_square.get_piece();
    
    if (!the_piece) {
        throw std::runtime_error("No piece at source position");
    }
    
    the_piece->set_position(to_pos);
    Square& new_square = get_square(to_pos);
    std::unique_ptr<Piece> taken = new_square.set_piece(the_piece);
    
    if (taken) {
        // Захват фигуры
        remove_piece(taken);
        save_UMC = uncapturing_moves_count;
        uncapturing_moves_count = 0;
    } else {
        ++uncapturing_moves_count;
    }
    
    old_square.remove_piece();
    
    if (returning) {
        if (!history.empty()) history.pop_back();
        --uncapturing_moves_count;
        --uncapturing_moves_count;
    } else {
        history.emplace_back(std::pair(from_pos, to_pos));
    }
    
    if (set_instead) {
        add_piece(set_instead);
        uncapturing_moves_count = save_UMC;
    }
    
    return taken;
}

std::pair<std::unique_ptr<Piece>, Vector> Board::is_in_check(Color color) {
    
    std::unique_ptr<Piece> gen_red = nullptr;
    std::unique_ptr<Piece> gen_black = nullptr;
    
    for (const auto& piece : reds) {
        if (dynamic_cast<General*>(piece.get())) {
            gen_red = piece;
            break;
        }
    }
    
    for (const auto& piece : blacks) {
        if (dynamic_cast<General*>(piece.get())) {
            gen_black = piece;
            break;
        }
    }
    
    if (gen_red && gen_black) {
        int x = gen_red->get_position().x;
        Vector red_pos = gen_red->get_position();
        Vector black_pos = gen_black->get_position();
        
        if (x == black_pos.x) {
            int y1 = red_pos.y;
            int y2 = black_pos.y;
            
            if (y1 > y2) std::swap(y1, y2);
            
            if (empty_between_general(x, y1, y2)) {
                if (color == Color::BLACK) {
                    return {gen_red, Vector(x, 10)};
                }
                else {
                    return {gen_black, Vector(x, 10)};
                }
            }
        }
    }
    
    // Проверка повторения позиции (правило трехкратного повторения)
    if (history.size() >= 8) {
        bool repeated3 = true;
        for (int i = 0; i < 4; i++) {
            const auto& move1 = history[history.size() - 1 - i];
            const auto& move2 = history[history.size() - 5 - i];
            
            if (move1 != move2) {
                repeated3 = false;
                break;
            }
        }
        
        if (repeated3) {
            //std::cout << "REPEATED" << std::endl;
            const auto& last = history.back();
            return {get_square(last.second).get_piece(), last.second};
        }
    }
    
    // Проверка атак других фигур
    const auto& opponent_pieces = (color == Color::BLACK) ? reds : blacks;
    const auto& palace = (color == Color::RED) ? RED_PALACE_AREA : BLACK_PALACE_AREA;
    
    for (const auto& attacker : opponent_pieces) {
        // Проверяем только атакующие фигуры
        if (!attacker->is_attacker()) continue;
        
        // Получаем возможные ходы без проверки на шах
        std::vector<Vector> attackers_valid_moves = get_piece_valid_moves(
            attacker.get(), 
            false,   // check = false
            false    // for_eval = false
        );
        
        for (const auto& move : attackers_valid_moves) {
            try {
                Square& square = get_square(move, palace);
                auto piece = square.get_piece();
                
                // Проверяем, стоит ли на клетке генерал
                if (piece && dynamic_cast<General*>(piece.get()) && 
                    piece->get_color() == color) {
                    return {attacker, move};
                }
            } catch (const std::out_of_range&) {
                // Пропускаем ходы за пределы дворца
                continue;
            }
        }
    }
    
    // Шах не обнаружен
    return {nullptr, Vector()};
}

std::vector<Vector> Board::get_piece_valid_moves(Piece* piece, bool check = true, bool for_eval = false) {
    std::vector<Vector> valid_moves;
    bool was_modified = false;

    // Модификация солдата при пересечении реки
    if (piece->can_modify() && !piece->is_modified()) {
        int y = piece->get_position().y;
        if ((piece->get_color() == Color::RED && y > 4) || 
            (piece->get_color() == Color::BLACK && y < 5)) {
            piece->modify();
            was_modified = true;
        }
    }

    const Move& move = piece->get_move();
    Vector start_pos = piece->get_position();

    for (const Vector& direction : move.get_directions()) {
        // Проверка для прыгающих фигур (конь, слон)
        if (move.is_bigstep()) {
            Vector midpoint = start_pos + (direction / 2);
            if (has_piece(midpoint)) {
                continue;
            }
        }

        bool has_screen = false;
        Vector current_pos = start_pos;
        int max_steps = piece->get_max_steps();

        for (int step = 0; step < max_steps; step++) {
            current_pos = current_pos + direction;
            
            // Проверка выхода за допустимую область
            try {
                Square& target_square = get_square(current_pos, piece->get_area());
            } catch (const std::out_of_range&) {
                break;
            }

            Square& target_square = get_square(current_pos);
            
            if (target_square.get_piece()) {
                // Обработка захвата фигуры
                if (!piece->need_screen() || has_screen) {
                    // Можно атаковать только вражеские фигуры
                    if (for_eval || target_square.get_piece()->get_color() != piece->get_color()) {
                        if (check) {
                            // Проверка на шах
                            auto [threat, killer_move] = ghost_test(
                                start_pos, current_pos,
                                [&](Board& b) {
                                    return b.is_in_check(piece->get_color());
                                }
                            );
                            
                            if (threat) {
                                if (debug) {
                                    std::cout << "Move by " << piece->get_name() 
                                                << " to " << current_pos 
                                                << " causes mate by " << threat->get_name()
                                                << " with move to " << killer_move << std::endl;
                                }
                                break;
                            }
                        }
                        valid_moves.push_back(current_pos);
                    }
                    break;
                } else {
                    // Фигура становится экраном для пушки
                    has_screen = true;
                }
            } else {
                // Обработка перемещения на пустую клетку
                if (for_eval ? (has_screen == piece->need_screen()) : !has_screen) {
                    if (check) {
                        // Проверка на шах
                        auto [threat, killer_move] = ghost_test(
                            start_pos, current_pos,
                            [&](Board& b) {
                                return b.is_in_check(piece->get_color());
                            }
                        );
                        
                        if (!threat) {
                            valid_moves.push_back(current_pos);
                        } else if (debug) {
                            std::cout << "BOT Move by " << piece->get_name() 
                                        << " to " << current_pos 
                                        << " causes mate by " << threat->get_name()
                                        << " with move to " << killer_move << std::endl;
                        }
                    } else {
                        valid_moves.push_back(current_pos);
                    }
                }
            }
        }
    }

    // Восстановление исходного состояния солдата
    if (piece->can_modify() && piece->is_modified()) {
        piece->set_default();
    }

    return valid_moves;
}

bool Board::has_piece(const Vector& position){
    try {
        Square& sq = get_square(position);
        return static_cast<bool>(sq.get_piece());
    } catch (const std::out_of_range&) {
        return false;
    }
}

double Board::is_mate() const {
    int mate = 3;
    for (int i = 0; i < 2; i++) {
        const auto& team = (i == 0) ? reds : blacks;
        for (const auto& piece : team) {
            if (dynamic_cast<General*>(piece.get())) {
                mate -= i + 1;
                break;
            }
        }
    }
    if (mate == 0) return 0;
    return (mate == 1) ? -std::numeric_limits<double>::infinity()
                        : std::numeric_limits<double>::infinity();
}

double Board::evaluate(const EvaluateSet& eval_set, bool describe = false) {
    if (uncapturing_moves_count > 49) return 0;
    
    double mate_result = is_mate();
    if (std::isinf(mate_result)) return mate_result;
    
    reset_control_state();
    auto [value_multiplier, attack_bonus, mobility_multiplier, control_multiplier] = eval_set.get();
    double total_value = 0;
    
    for (int i = 0; i < 2; i++) {
        Color color = (i == 0) ? Color::RED : Color::BLACK;
        const auto& team = (i == 0) ? reds : blacks;
        
        double team_values = 0;
        double team_attack_bonus = 0;
        double team_mobility = 0;
        std::vector<Vector> vds; // valid moves
        
        for (const auto& piece : team) {
            team_values += piece->get_value() * value_multiplier;
            
            if (attack_bonus && piece->is_attacker()) {
                team_attack_bonus += attack_bonus;
            }
            
            if (mobility_multiplier || control_multiplier) {
                // Заглушка - предполагается реализация get_piece_valid_moves
                vds = get_piece_valid_moves(piece.get(), true, true);
                team_mobility += vds.size() * mobility_multiplier;
            }
            
            if (control_multiplier) {
                update_control_state(color, vds);
            }
            
            if (describe) {
                std::cout << "BOARD: " << (color == Color::RED ? "RED" : "BLACK")
                            << " values = " << team_values << "\n";
                std::cout << "BOARD: " << (color == Color::RED ? "RED" : "BLACK")
                            << " attack bonus = " << team_attack_bonus << "\n";
                std::cout << "BOARD: " << (color == Color::RED ? "RED" : "BLACK")
                            << " mobility = " << team_mobility << "\n";
            }
        }
        
        double team_total = team_values + team_attack_bonus + team_mobility;
        total_value += (i == 0) ? team_total : -team_total;
    }
    
    if (control_multiplier) {
        for (auto& row : squares) {
            for (auto& square : row) {
                if (!square.get_piece()) {
                    double control_value = 0;
                    switch (square.get_state()) {
                        case Square_status::red:
                            control_value = control_multiplier;
                            break;
                        case Square_status::black:
                            control_value = -control_multiplier;
                            break;
                        default:
                            break;
                    }
                    total_value += control_value;
                }
            }
        }
    }
    
    return total_value;
}

double Board::update_evaluation(const EvaluateSet& eval_set, bool describe = false) {
    evaluation = evaluate(eval_set, describe);
    return evaluation;
}

void Board::update_control_state(Color color, const std::vector<Vector>& attacks) {
    for (const auto& attack : attacks) {
        try {
            Square& square = get_square(attack);
            square.add_attack(color);
        } catch (const std::out_of_range&) {
            // Игнорируем атаки за пределами доски
        }
    }
}

void Board::reset_control_state() {
    for (auto& row : squares) {
        for (auto& square : row) {
            square.reset_attacks();
        }
    }
}

void Board::print_visual(Color perspective = Color::RED){

    std::string line = " ╔";
    for (int i = 0; i < 9; ++i) {
        line += "═══";
    }
    line += "╗";
    std::cout << line << std::endl;

    // Основная часть доски (10 строк)
    for (int i = 0; i < 10; ++i) {
        int vert = (perspective == Color::BLACK) ? i : (9 - i);
        line = std::to_string(vert) + "║";

        for (int j = 0; j < 9; ++j) {
            const std::unique_ptr<Piece> piece = squares[j][vert].get_piece();
            std::string between_char;

            // Определяем символ-разделитель для сетки
            if (i == 4 || i == 9) {
                between_char = "┴";
            } else if (i == 0 || i == 5) {
                between_char = "┬";
            } else if (j == 0) {
                between_char = "├";
            } else if (j == 8) {
                between_char = "┤";
            } else {
                between_char = "┼";
            }

            if (piece != nullptr) {
                // Добавляем фигуру с цветом
                line += " ";
                line += (piece->get_color() == Color::RED) ? RED_TEXT : GREEN_TEXT;
                line += piece->get_name();
                line += RESET_TEXT;
                line += " ";
            } else {
                // Добавляем элементы сетки
                if (j == 0) {
                    line += " ";
                } else {
                    line += "─";
                }
                line += between_char;
                if (j == 8) {
                    line += " ";
                } else {
                    line += "─";
                }
            }
        }
        line += "║";
        std::cout << line << std::endl;
    }

    // Нижняя граница доски
    line = " ╚";
    for (int i = 0; i < 9; ++i) {
        line += "═══";
    }
    line += "╝";
    std::cout << line << std::endl;

    // Подпись столбцов
    line = "   ";
    for (int i = 0; i < 9; ++i) {
        line += std::to_string(i);
        line += "  ";
    }
    std::cout << line << std::endl << std::endl;
}

void Board::clear() {
    reds.clear();
    blacks.clear();
    
    history.clear();
    
    uncapturing_moves_count = 0;
    save_UMC = 0;
    evaluation = 0;
    
    squares = std::vector<std::vector<Square>>(9, std::vector<Square>(10, Square(Vector(0, 0))));
    for (int x = 0; x < 9; ++x) {
        for (int y = 0; y < 10; ++y) {
            squares[x][y] = Square(Vector(x, y));
        }
    }
}


#endif